README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# getrect

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg)
<!-- badges: end -->

`{getrect}` is a solution to the problem of partitioning a matrix into
rectangular regions - each region only enclosing a single unique value.

## What’s in the box

- `partition_matrix()` partition a matrix into rectangles
- `plot_rects()` Plot the result of `partition_matrix()`
- `max_rect_under_histogram()` calculates the maximum area rectangle
  under a sequence of values representing bar heights of a histogram
- `plot_hist_max_rect()` plot the bar heights and best rectangle

## Future ideas

- There are faster algorithms for ‘max area under a histogram’ than what
  I’ve included here.
- When only rectangles with area = 1 remain, could short-circuit the
  iteration and add them all in bulk.

## Installation

You can install from [GitHub](https://github.com/coolbutuseless/getrect)
with:

``` r
# install.package('remotes')
remotes::install_github('coolbutuseless/getrect')
```

# Problem: Partition matrix into rectangular regions

Partition a matrix into rectangular regions where:

- Each region contains only a single unique value
- An effort is made to minimise the total number of rectangles

# Solution:

- For each unique value:
  - Greedily find the biggest single-valued rectangle at each step

## Key algorithm: Maximum rectangular area under a histogram

Given a sequence of bar heights (representing a histogram), what is the
maximum sized rectangle which can be inscribed?

``` r
library(getrect)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Define bar heights
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
hts <- c(4, 8, 3, 1, 3, 3, 0, 1)
plot_hist_max_rect(hts)
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Calculate maximum sized rectangle
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
rect <- max_rect_under_histogram(hts)
plot_hist_max_rect(hts, rect)
```

<img src="man/figures/README-example-2.png" width="100%" />

# Step-by-step example on example matrix

``` r
mat <- matrix(
  c(1,1,1,2,3,3,3,
    1,1,2,2,2,3,3,
    1,2,2,2,2,2,3,
    2,2,2,2,2,2,2), 
  nrow = 4, ncol = 7, byrow = TRUE
)
mat
#>      [,1] [,2] [,3] [,4] [,5] [,6] [,7]
#> [1,]    1    1    1    2    3    3    3
#> [2,]    1    1    2    2    2    3    3
#> [3,]    1    2    2    2    2    2    3
#> [4,]    2    2    2    2    2    2    2
```

## Consider a single target value

Consider just the value `2` in the matrix.

Create a working matrix `s` with a `1` where the target value exists,
and `0` elsewhere.

``` r
target <- 2
s <- mat
s[] <- mat == 2
s
#>      [,1] [,2] [,3] [,4] [,5] [,6] [,7]
#> [1,]    0    0    0    1    0    0    0
#> [2,]    0    0    1    1    1    0    0
#> [3,]    0    1    1    1    1    1    0
#> [4,]    1    1    1    1    1    1    1
```

## Calculate the vertical runs

Calculate the vertical runs of `1` at every row based upon the rows
above

``` r
s <- apply(s, 2, \(x) ave(x, cumsum(x == 0), FUN = cumsum))
s
#>      [,1] [,2] [,3] [,4] [,5] [,6] [,7]
#> [1,]    0    0    0    1    0    0    0
#> [2,]    0    0    1    2    1    0    0
#> [3,]    0    1    2    3    2    1    0
#> [4,]    1    2    3    4    3    2    1
```

## Consider each row in isolation, and calculate the *max rect under the histogram*

``` r
row_max_rects <- apply(s, 1, \(x) as.data.frame(max_rect_under_histogram(x)))
row_max_rects <- do.call(rbind, row_max_rects)
row_max_rects
#>   area area_alt  ar ar_alt xmin xmax h w
#> 1    1        1 1.0    1.0    4    4 1 1
#> 2    3        3 3.0    3.0    3    5 1 3
#> 3    6        6 1.5    1.5    3    5 2 3
#> 4   10       10 2.5    2.5    2    6 2 5
```

## Pick the best row (max rect area)

``` r
idx <- which.max(row_max_rects$area)
best_rect <- row_max_rects[idx, ]

# Calculate the y extents of the rectangle
best_rect$ymax <- idx
best_rect$ymin <- idx - best_rect$h + 1
best_rect$target <- target
```

## Mark part of matrix as done

``` r
plot_rects(best_rect, mat)
```

<img src="man/figures/README-unnamed-chunk-7-1.png" width="100%" />

## Repeat until no unmarked areas remain

<img src="man/figures/README-unnamed-chunk-8-1.png" width="100%" />

# Larger example

``` r
set.seed(2)
w <- 15
h <- 15
mat <- matrix(sample(c(1, 2), w*h, replace = TRUE, prob = c(4, 1)), h, w)
mat
#>       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
#>  [1,]    1    2    1    1    1    1    2    1    1     1     1     1     2
#>  [2,]    1    2    1    2    2    1    1    2    1     1     1     1     1
#>  [3,]    1    1    2    1    1    2    1    1    2     2     2     2     2
#>  [4,]    1    1    2    1    1    1    1    1    2     1     1     1     1
#>  [5,]    2    1    1    2    2    1    1    1    1     1     1     1     1
#>  [6,]    2    1    1    1    1    1    1    1    1     1     1     2     2
#>  [7,]    1    1    2    1    1    1    1    1    1     1     2     1     1
#>  [8,]    2    2    1    1    1    1    1    1    1     1     1     1     1
#>  [9,]    1    1    1    2    1    1    1    1    1     1     1     2     1
#> [10,]    1    1    1    1    1    1    1    2    1     1     1     1     1
#> [11,]    1    1    2    2    1    2    1    2    1     1     2     1     1
#> [12,]    1    1    1    1    1    2    1    2    2     1     2     1     1
#> [13,]    1    1    1    2    1    1    2    1    1     1     1     1     2
#> [14,]    1    2    1    1    1    1    2    2    2     1     1     1     1
#> [15,]    1    1    2    1    1    1    1    1    1     1     1     1     1
#>       [,14] [,15]
#>  [1,]     1     2
#>  [2,]     1     1
#>  [3,]     1     2
#>  [4,]     1     1
#>  [5,]     1     2
#>  [6,]     2     1
#>  [7,]     1     1
#>  [8,]     1     1
#>  [9,]     2     1
#> [10,]     1     1
#> [11,]     2     1
#> [12,]     1     2
#> [13,]     1     1
#> [14,]     1     1
#> [15,]     1     1

# Partition
rects <- partition_matrix(mat)
plot_rects(rects)
```

<img src="man/figures/README-unnamed-chunk-9-1.png" width="100%" />

# Extension: prefer areas with an aspect ratio close to 1

Same matrix as above, but penalise the selection of rectangles with high
aspect ratio by setting `ar_penalty`.

``` r
rects <- partition_matrix(mat, ar_penalty = 0.9)
plot_rects(rects)
```

<img src="man/figures/README-unnamed-chunk-10-1.png" width="100%" />


NEWS.md
# getrect 0.1.0  2023-10-21

* Initial release


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(getrect)

test_check("getrect")


tests/testthat/test-get-rect.R



test_that("multiplication works", {
  mat <- matrix(
    c(1,1,1,2,3,3,3,
      1,1,2,2,2,3,3,
      1,2,2,2,2,2,3,
      2,2,2,2,2,2,2), 
    nrow = 4, ncol = 7, byrow = TRUE
  )
  
  rects <- partition_matrix(mat, ar_penalty = 0)
  # plot_rects(rects)
  
  expect_equal(
    rects$area, 
    c(4, 1, 1, 10, 3, 1, 1, 1, 4, 1, 1)
  )
  
})


tests/testthat/test-max-rect.R

test_that("max-rect-under-histogram works", {

  
  hts <- c(6, 2, 5, 4, 5, 1, 6) # 12
  
  expect_equal(
    max_rect_under_histogram(hts)$area,
    12
  )

  hts <- c(4,8,3,1,1,0) # 9
  expect_equal(
    max_rect_under_histogram(hts)$area,
    9
  )
  
  hts <- c(2, 2, 1, 2, 2)
  expect_equal(
    max_rect_under_histogram(hts)$area,
    5
  )
})


test_that("ar_penalty works", {
  hts <- c(2, 2, 1, 2, 2)
  expect_equal(
    max_rect_under_histogram(hts, ar_penalty = 1)$area,
    4
  )
})


R/max-rectangle.R

globalVariables(c('x', 'y'))

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Find the area of the maximum rectangle under the given histogram
#'
# This implementation is pretty brute force, but makes use of rle() to 
# find matching bars for the current bar, so it was easy to write at least :)
#'
#' Having the guts exposed in this function also help in manually tweaking the 
#' accepted "maximum area rectangle" by penalising areas we don't want.
#' Currently high-aspect-ratio areas are penalised so that areas which 
#' are more square might be chosen (even if actual total area is less)
#' 
#' @param hts histogram bar heights e.g. \code{c(2, 2, 1, 2, 2)}
#' @param ar_penalty Aspect ratio penalty (multiplication factor). Default value
#'        of \code{0} means to return the largest area regardless of aspect ratio.   
#'        Values above 0 are used to reduce the area by this factor multiplied
#'        by the aspect ratio of the rectangle.
#'        Value can be any 
#'        positive floating point number. Useful range: 0-20
#'
#' @export
#' 
#' @examples
#' \dontrun{
#' hts <- c(2, 2, 1, 2, 2)
#' plot_hist_max_rect(hts, rect = NULL)
#' rect <- max_rect_under_histogram(hts)
#' plot_hist_max_rect(hts, rect = rect)
#' 
#' rect <- max_rect_under_histogram(hts, ar_penalty = 1)
#' plot_hist_max_rect(hts, rect = rect)
#' }
#' 
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
max_rect_under_histogram <- function(hts, ar_penalty = 0) {
  
  res <- list(
    area = -Inf
  )
  
  # Short circuit if nothing to do
  if (all(hts <= 0)) {
    return(list(
      area     = 0,
      area_alt = 0, #area_alt,
      ar       = 9, #ar,
      ar_alt   = 9, #ar_alt,
      xmin     = 1, #x_range[1],
      xmax     = 1, #x_range[2],
      h        = 0, #hts[idx],
      w        = 0  #width,
    ))
  }
  
  # for each bar in the histogram
  #   - find matching bars which as this bar height or above
  #   - keep only the bars of matching height which are 
  #     part of a run of bars which include the current bar
  #   - area = height of bar * length of run of matching bars
  for (idx in seq_along(hts)) {
    
    # SOA = Same height Or Above (compared to bar hts[idx])
    ht  <- hts[idx]
    if (ht == 0) next
    soa <- hts >= ht
    
    # Run-length encode the matching bars
    rl <- unclass(rle(soa))
    
    # where to these 'runs' start
    starts <- c(1, head(1 + cumsum(rl$lengths), -1))
    
    # find the 'run' which includes the current bar
    rl_idx <- max(which(idx >= starts))
    
    # how many peers are in this same 'run'?
    # i.e. bars which are same height or above.
    npeers <- rl$lengths[[rl_idx]]
    
    # what are the actual indices of these peers?
    peer_idxs <- starts[rl_idx] + seq_len(npeers) - 1
    x_range   <- range(peer_idxs)
    
    # Total area of this rectangle
    #     = width  * height
    area <- npeers * ht
    
    # Aspect ratio
    width <- diff(x_range) + 1
    ar_alt <- ar <- width/ht
    if (ar_alt < 1) {
      ar_alt <- 1/ar_alt
    }
    
    # Penalise the area by a multiple of its aspect ratio.
    # i.e. Prefer "squarer" areas by artificially reducing
    # the calcualted area (area_alt) by a multiple of the
    # aspect ratio.  
    # Ingore aspect ratio for area = 1 or 2.
    if (area < 3) {
      area_alt <- area
    } else {
      area_alt <- area - ar_penalty * ar_alt
    }
    
    
    if (area_alt > res$area) {
      res$area     <- area
      res$area_alt <- area_alt
      res$ar       <- ar
      res$ar_alt   <- ar_alt
      res$xmin     <- x_range[1]
      res$xmax     <- x_range[2]
      res$h        <- hts[idx]
      res$w        <- width
    }
    
  }
  res
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Plot a histogram and its max rectangle
#' 
#' @param hts bar heights of histogram
#' @param rect output from \code{max_rect_under_histogram()}
#' 
#' @return ggplot object
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
plot_hist_max_rect <- function(hts, rect = NULL) {
  
  p <- NULL
  
  if (requireNamespace('ggplot2', quietly = TRUE)) {
    plot_df <- data.frame(x = seq_along(hts), y = hts)
    
    p <- ggplot2::ggplot(plot_df) + 
      ggplot2::geom_col(ggplot2::aes(x, y), width = 1, fill = 'grey80', colour = 'grey30') + 
      ggplot2::theme_minimal() 
    
    if (!is.null(rect)) {
      p <- p + ggplot2::annotate(
        'rect', 
        xmin     = rect$xmin - 0.5, 
        xmax     = rect$xmax + 0.5,
        ymin     = 0,
        ymax     = rect$h,
        fill     = 'dodgerblue3',
        alpha    = 0.3,
        colour   = 'black', 
        linetype = 2
      )
    }
  } else {
    stop("{ggplot2} package needed for plotting")
  }
  
  p
}



if (FALSE) {
  hts <- c(6, 2, 5, 4, 5, 1, 6) # 12
  max_rect <- max_rect_under_histogram(hts) 
  plot_hist_max_rect(hts, max_rect)
  
  plot_rect(max_rect)
}



R/partition.R

globalVariables(c('area', 'area_alt', 'aspect', 'xmin', 'xmax', 'ymin', 'ymax',
                  'target', 'val'))


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Create a data.frame of rectangular areas of the target value in the matrix
#' 
#' @inheritParams max_rect_under_histogram
#' @param mat matrix
#' @param target target value
#'
#' @importFrom stats ave
#' @importFrom utils head
#'
#' @return data.frame
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
partition_matrix_single <- function(mat, target, ar_penalty = 0) {
  
  max_areas <- list()
  
  while (TRUE) {
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # 0/1 matrix indicating where the target value is.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    s   <- mat
    s[] <- as.integer(s == target)
    mode(s) <- 'integer'
    
    if (all(as.vector(s) == 0)) {
      # print("No target values left")
      break
    }
    
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Cumulative counts of the target value vertically.
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    s <- apply(s, 2, \(x) ave(x, cumsum(x == 0), FUN = cumsum))
    
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Find largest rectangle by
    #   * find the 'max area under histogram' for each row
    #   * select the row with the largest 'max area under histogram'
    #   * if max area is tied, prefer areas with aspect ratio closest to 1
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    areas <- lapply(
      seq(nrow(s)), 
      function(i) {
        # print(i)
        res <- max_rect_under_histogram(s[i,], ar_penalty = ar_penalty)
        res$ymin <- i - res$h + 1
        res$ymax <- i
        if (res$area < 0) {
          NULL
        } else {
          as.data.frame(res)
        }
      })
    
    
    area_df <- do.call(rbind, areas)
    
    
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Sort the area by some criteria and then pick the top one
    #    * pick largest area (greedy!)
    #    * if multiple areas of the same maximum size
    #         * choose area with aspect ratio closest to 1
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    area_df <- subset(area_df, area > 0)
    area_df <- area_df[with(area_df, order(-area_alt, ar_alt)),]
    
    if (nrow(area_df) == 0) {
      # zero areas can happen if all row is "-Inf"
      # print("No non-zero area")
      break;
    }
    
    # Note: If all areas == 1, then can probably add them in one go
    # rather than 1-at-a-time

    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Best area is at the top
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    max_area <- area_df[1,]
    
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Add the max_area from this iteration to the list of all max areas
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    max_areas <- c(max_areas, list(max_area))
    
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    # Change the values of current max area in the orignal matrix to -Inf 
    # This will mark it as "taken" so that the next iteration will ignore them
    #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    mat[(max_area$ymin):(max_area$ymax), (max_area$xmin):(max_area$xmax)] <- -Inf
    
  }
  
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # All areas for this target
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  max_areas <- do.call(rbind, max_areas)
  max_areas$target <- target
  
  max_areas
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Determine all rectangular regions of the same value in the matrix
#' 
#' @inheritParams max_rect_under_histogram
#' @param mat matrix
#' 
#' @return data.frame
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
partition_matrix <- function(mat, ar_penalty = 0) {
  
  rects <- lapply(
    unique(as.vector(mat)), 
    \(target) partition_matrix_single(mat, target = target, ar_penalty = ar_penalty)
  )
  
  do.call(rbind, rects)
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Plot all rects in a matrix
#'
#' @param rects rects returned by \code{partition_matrix()}
#' @param mat matrix
#'
#' @return ggplot object
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
plot_rects <- function(rects, mat = NULL) {
  
  p <- NULL
  
  if (requireNamespace('ggplot2', quietly = TRUE)) {
    p <- ggplot2::ggplot(rects) +
      ggplot2::geom_rect(
        mapping = ggplot2::aes(
          xmin = xmin - 1, xmax = xmax, ymin = ymin - 1, ymax = ymax,
          fill = as.factor(target)
        ),
        alpha = 0.5,
        colour = 'black'
      ) + 
      ggplot2::scale_y_reverse() +
      ggplot2::coord_equal() +
      ggplot2::theme_void() 
    
    if (!is.null(mat)) {
      text_df <- data.frame(
        x   = rep(seq(ncol(mat)), each = nrow(mat)),
        y   = rep(seq(nrow(mat)), ncol(mat)),
        val = as.character(as.vector(mat))
      )
      
      p <- p + 
        ggplot2::geom_text(
          data = text_df, 
          mapping = ggplot2::aes(x - 0.5, y - 0.5, label = val)
        ) + 
        ggplot2::theme(legend.position = 'none')
    } else {
      p <- p + ggplot2::theme(legend.position = 'bottom')
    }
    
    p
  } else {
    stop("{ggplot2} package needed for plotting")
  }
  
  p
}



if (FALSE) {
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Create a random matrix
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  set.seed(2)
  w <- 15
  h <- 15
  mat <- matrix(sample(c(1, 2), w*h, replace = TRUE, prob = c(4, 1)), h, w)
  
  rects <- partition_matrix(mat, ar_penalty = 1)
  
  plot_rects(rects, mat)
  
}
